src/Storages/MergeTree/MergeTreeIndexMinMax.cpp

```diff
@@ -54,14 +71,30 @@ void MergeTreeIndexGranuleMinMax::deserializeBinary(ReadBuffer & istr)
     {
         const DataTypePtr & type = index_sample_block.getByPosition(i).type;
         auto serialization = type->getDefaultSerialization();
-        serialization->deserializeBinary(min_val, istr);
-        serialization->deserializeBinary(max_val, istr);
-
-        // NULL_LAST
-        if (min_val.isNull())
-            min_val = PositiveInfinity();
-        if (max_val.isNull())
-            max_val = PositiveInfinity();
+
+        if (!type->isNullable())
+        {
+            serialization->deserializeBinary(min_val, istr);
+            serialization->deserializeBinary(max_val, istr);
+        }
+        else
+        {
+            /// NOTE: that this serialization differs from
+            /// IMergeTreeDataPart::MinMaxIndex::load() due to preserve
+            /// backward compatibility.
+            bool is_null;
+            readBinary(is_null, istr);
+            if (!is_null)
+            {
+                serialization->deserializeBinary(min_val, istr);
+                serialization->deserializeBinary(max_val, istr);
+            }
+            else
+            {
+                min_val = Null();
+                max_val = Null();
```

Comment:
Let's use `PositiveInfinity` here for both endpoints.

Reply:
This cannot be done since in this case it will try to filter granulas even if it has at least one NULL value (IIUC)

Reply:
Indeed. We didn't support Null index at all.